### PR TITLE
Manage subdirectories in submissions file manager

### PR DESCRIPTION
**Pull request by mishaschwartz**
- enables creation and deletion of subdirectories
- enables uploading files to subdirectories
- only enabled for the submission_file_manager (others will come in a future PR)
- introduces a new react modal for uploading files
- introduces a new policy to check whether a user can create, remove or change subdirectories
- introduces new props to the raw file manager component that selectively prevents certain action bar elements from being created

### CODE REVIEW

File: app/assets/javascripts/Components/Modals/submission_file_upload_modal.jsx [hunk 1]
```diff
@@ -0,0 +1,49 @@
+import React from 'react';
+import Modal from 'react-modal';
+
+class SubmissionFileUploadModal extends React.Component {
+
+  constructor(props) {
+    super(props);
+    this.state = {
+      newfiles: []
```
Comment (david-yz-liu) [inline app/assets/javascripts/Components/Modals/submission_file_upload_modal.jsx]: Let's use camel case here (`newFiles`).

File: app/assets/javascripts/Components/Modals/submission_file_upload_modal.jsx [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    }
```
Comment (david-yz-liu) [inline app/assets/javascripts/Components/Modals/submission_file_upload_modal.jsx]: Please put a semicolon here.

File: app/assets/javascripts/Components/Modals/submission_file_upload_modal.jsx [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+  };
+
+  render() {
+    return (
+      <Modal
+        className="react-modal"
+        isOpen={this.props.isOpen}
+        onRequestClose={this.props.onRequestClose}
+      >
+        <h2>{I18n.t('add_new')}</h2>
```
Comment (david-yz-liu) [inline app/assets/javascripts/Components/Modals/submission_file_upload_modal.jsx]: Oh man let's use this opportunity to get rid of this misnamed key. I'd be happy with `I18n.t('upload')` ("Upload"), or you can use `'upload_the'` and create a new generic key for `"File"` if you want.

File: app/assets/javascripts/Components/Modals/submission_file_upload_modal.jsx [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+        <form onSubmit={this.onSubmit}>
+          <div className={'modal-container-vertical'}>
+            <div className={'modal-container'}>
+              <input type={'file'} name={'newfiles'} multiple={true} onChange={this.handleFileUpload}/>
```
Comment (david-yz-liu) [inline app/assets/javascripts/Components/Modals/submission_file_upload_modal.jsx]: The "newfiles" here is a bit tricky. I guess to make the name realistic you can use `'new_files'` (this is the name of the param that's sent back).

File: app/assets/javascripts/Components/markus_file_manager.jsx [hunk 5]
```diff
@@ -10,13 +10,21 @@ import { RawFileBrowser, Headers, FileRenderers, BaseFileConnectors } from 'reac
 
 
 class RawFileManager extends RawFileBrowser {
-  handleActionBarAddFileClick = (event) => {
+  handleActionBarAddFileClick = (event, selected) => {
     event.preventDefault();
-    window.modal_addnew.open();
+    let uploadTarget = '';
+    if (selected) {
+      if (selected.children) {
```
Comment (david-yz-liu) [inline app/assets/javascripts/Components/markus_file_manager.jsx]: Does this work if a folder is selected, but it has no children? (Is `selected.children` an empty list in this case?)

I believe the react-keyed-file-browser way of detecting whether something's a folder or not is whether its name ends with a `/`. Kind of weird, but I think maybe we should just be consistent with that.

File: app/assets/javascripts/Components/markus_file_manager.jsx [hunk 6]
```diff
@@ -10,13 +10,21 @@ import { RawFileBrowser, Headers, FileRenderers, BaseFileConnectors } from 'reac
...
+        uploadTarget = selected.relativeKey;
+      } else {
+        uploadTarget = selected.relativeKey.substring(0, selected.relativeKey.lastIndexOf(selected.name));
+      }
+    }
+    this.props.onActionBarAddFileClick(uploadTarget)
```
Comment (david-yz-liu) [inline app/assets/javascripts/Components/markus_file_manager.jsx]: Please put a semicolon here.

File: app/assets/javascripts/Components/markus_file_manager.jsx [hunk 7]
```diff
@@ -10,13 +10,21 @@ import { RawFileBrowser, Headers, FileRenderers, BaseFileConnectors } from 'reac
...
   };
 
   renderActionBar(selectedItem) {
-    const selectionIsFolder = selectedItem && !selectedItem.size;
+    const selectionIsFolder = selectedItem && selectedItem.children;
```
Comment (david-yz-liu) [inline app/assets/javascripts/Components/markus_file_manager.jsx]: See above comment.

File: app/assets/javascripts/Components/submission_file_manager.jsx [hunk 8]
```diff
@@ -1,14 +1,16 @@
...
 
 
 class SubmissionFileManager extends React.Component {
 
   constructor(props) {
     super(props);
     this.state = {
-      files: []
+      files: [],
+      showModal: false
```
Comment (david-yz-liu) [inline app/assets/javascripts/Components/submission_file_manager.jsx]: Seems like you need to add an `uploadTarget` state as well?

File: app/assets/javascripts/Components/submission_file_manager.jsx [hunk 9]
```diff
@@ -105,17 +134,32 @@ class SubmissionFileManager extends React.Component {
     });
   };
 
+  uploadFiles = (uploadTarget) => {
```
Comment (david-yz-liu) [inline app/assets/javascripts/Components/submission_file_manager.jsx]: I think this name is a bit misleading. You could do something like `openUploadModal`.

File: app/controllers/submissions_controller.rb [hunk 10]
```diff
@@ -684,18 +702,22 @@ def get_all_file_data(revision, grouping, path)
...
+        dirname, basename = File.split(file_name)
+        dirname = '' if dirname == '.'
+        data = get_file_info(basename, file_obj, grouping.assignment.id,
+                             revision.revision_identifier, dirname, grouping.id)
+        next if data.nil?
+        data[:key] = file_name
+        data[:modified] = data[:last_revised_date]
+        data
+      else
+        {key: "#{file_name}/"}
```
Comment (david-yz-liu) [inline app/controllers/submissions_controller.rb]: Put spaces inside the `{` and `}` (not sure why Hound didn't complain?)

File: app/assets/javascripts/Components/markus_file_manager.jsx (hunk 5 above)
Comment (mishaschwartz) [reply app/assets/javascripts/Components/markus_file_manager.jsx]: An empty directory still has a `children` attribute but I agree that keeping it consistent with the package behaviour makes sense.

File: app/assets/javascripts/Components/submission_file_manager.jsx (hunk 8 above)
Comment (mishaschwartz) [reply app/assets/javascripts/Components/submission_file_manager.jsx]: it's not necessary but it's good to add so the reader is aware of all possible state keys.

File: app/assets/javascripts/Components/markus_file_manager.jsx [hunk 11]
```diff
@@ -57,27 +81,30 @@ class RawFileManager extends RawFileBrowser {
...
             <li key="action-add-folder">
               <a
-                onClick={this.handleActionBarAddFolderClick}
+                onClick={(event) => this.handleActionBarAddFolderClickSetSelection(event, target)}
                 href="#"
                 role="button"
               >
                 <i className="fa fa-folder-o" aria-hidden="true"/>
-                &nbsp;Add Subfolder
+                &nbsp;{target ? "Add Subolder" : "Add Folder"}
```
Comment (david-yz-liu) [inline app/assets/javascripts/Components/markus_file_manager.jsx]: There's a typo on "Subolder"

**Pull request merged**